petastorm/pytorch.py

```diff
@@ -339,16 +372,33 @@ def _iter_impl(self):
             # _yield_batches will emit as much batches as are allowed by the shuffling_buffer (RandomShufflingBuffer
             # will avoid underflowing below a certain number of samples to guarantee some samples decorrelation)
             for batch in self._yield_batches(keys):
+                if self.inmemory_cache_all:
+                    self._other_shuffling_buffer.add_many(batch.values())
                 yield batch
 
-        # Once reader can not read new rows, we might still have a bunch of rows waiting in the shuffling buffer.
-        # Telling shuffling buffer that we are finished allows to deplete the buffer completely, regardless its
+        # When caching memory, at this point we are done with reading all the data. Rest of the
+        # batches will be created directly from memory.
+        # Otherwise if there is not in memory cache, Once reader can not read new rows, we might
+        # still have a bunch of rows waiting in the shuffling buffer. Telling shuffling buffer
+        # that we are finished allows to deplete the buffer completely, regardless its
         # min_after_dequeue setting.
         self._shuffling_buffer.finish()
 
         for batch in self._yield_batches(keys):
+            if self.inmemory_cache_all:
+                self._other_shuffling_buffer.add_many(batch.values())
             yield batch
 
+        if self.inmemory_cache_all:
+            for _ in range(self.num_epochs - 1 if self.num_epochs else sys.maxsize):
+                self._other_shuffling_buffer.finish()
+                self._shuffling_buffer = self._other_shuffling_buffer
+                self._other_shuffling_buffer = instantiate_buffer_fn()
```

Comment:
Can self._other_shuffling_buffer be just local variables and not part of object state (we should probably also make self._shuffling_buffer a local variable, but it was like this before, so should not change in this PR)?